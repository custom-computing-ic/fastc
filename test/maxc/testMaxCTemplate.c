typedef int stream_t;
typedef int scalar_array;
typedef int count_t;

//void input(stream_t *x, int dim) {}
void output(stream_t* out, int res, int cond) {}
void chain(count_t i, count_t j, count_t k) {}
int fselect(int cond, int val1, int val2) {return (cond ==0? val1 : val2);}
count_t count(int wrap, int inc) {return 1;}
count_t countChain(int wrap, int inc, count_t parent) {return 1;}
int scalar(){return 1;}
//void input_scalar_array(scalar_array *c, int count) {}

#pragma in x CPU kernel_KernelFunc
#pragma out res CPU kernel_KernelFunc
#pragma scalar in c float8_24 kernel_KernelFunc
#pragma scalar in a1 uint32 kernel_KernelFunc
#pragma scalar in a2 uint32 kernel_KernelFunc
#pragma scalar in a3 uint32 kernel_KernelFunc
#pragma scalar in n1 uint32 kernel_KernelFunc
#pragma scalar in n2 uint32 kernel_KernelFunc
#pragma scalar in n3 uint32 kernel_KernelFunc
#pragma scalar in nx uint32 kernel_KernelFunc
#pragma scalar in ny uint32 kernel_KernelFunc
#pragma scalar in nz uint32 kernel_KernelFunc
void kernel_KernelFunc(stream_t *x, stream_t *res,
		 int nx, int ny, int nz, 
		 int a1, int a2, int a3,
		 int n1, int n2, int n3, 
		 int c) {

  // counters are required to keep track of point being 
  // processed and determine if this is within boundaries
  count_t i = count(nz, 1);
  count_t j = countChain(ny, 1, i);
  count_t k = countChain(nx, 1, j);

  // keep track if we are inside boundaries
  int cond = (i >= a1 & i < n1) & (j >= a2 & j < n2) & (k >= a3 & k < n3);

  int func =
          c * x[0] + 
          c * x[1]      + c * x[-1] +
          c * x[-nx]    + c * x[nx] + 
          c * x[-nx*ny] + c * x[nx*ny];


  // if inside boundaries apply the stencil operator
  // otherwise assign a default value (0)
  // int r = fselect(cond, func, 0);

  output(res, func, cond);
}
