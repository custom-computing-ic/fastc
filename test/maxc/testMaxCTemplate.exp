package engine;

import com.maxeler.maxcompiler.v1.kernelcompiler.*;
import com.maxeler.maxcompiler.v1.kernelcompiler.types.base.HWVar;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.*;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.core.*    ;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.core.Count.*;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.KernelMath.*;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.core.Stream.OffsetExpr;
import com.maxeler.maxcompiler.v1.kernelcompiler.types.composite.*;
import com.maxeler.maxcompiler.v1.kernelcompiler.types.base.*;

public class KernelFunc extends Kernel {

public KernelFunc(KernelParameters parameters) {
super(parameters);
HWVar x =  io.input("x", hwFloat(8, 24));
HWVar c =  io.scalarInput("c", hwFloat(8, 24));
HWVar a1 =  io.scalarInput("a1", hwUInt(32));
HWVar a2 =  io.scalarInput("a2", hwUInt(32));
HWVar a3 =  io.scalarInput("a3", hwUInt(32));
HWVar n1 =  io.scalarInput("n1", hwUInt(32));
HWVar n2 =  io.scalarInput("n2", hwUInt(32));
HWVar n3 =  io.scalarInput("n3", hwUInt(32));
HWVar nx =  io.scalarInput("nx", hwUInt(32));
HWVar ny =  io.scalarInput("ny", hwUInt(32));
HWVar nz =  io.scalarInput("nz", hwUInt(32));
CounterChain chain_i = control.count.makeCounterChain();


HWVar i = chain_i.addCounter(nz, 1);
HWVar j = chain_i.addCounter(ny, 1);
HWVar k = chain_i.addCounter(nx, 1);
    HWVar cond;
(cond = ((((i >= a1) & (i < n1)) & ((j >= a2) & (j < n2))) & ((k >= a3) & (k < n3))));
HWVar func;
(func = (((((((c * x) + (c * stream.offset(x, 1))) + (c * stream.offset(x, constant.var(-1), -1024, 1024))) + (c * stream.offset(x, -nx, -1024, 1024))) + (c * stream.offset(x, nx, -1024, 1024))) + (c * stream.offset(x, (-nx * ny), -1024, 1024))) + (c * stream.offset(x, (nx * ny), -1024, 1024))));
io.output("res", func, hwFloat(8, 24), cond);
}
}
