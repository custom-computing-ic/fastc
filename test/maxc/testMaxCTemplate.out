package engine;

import com.maxeler.maxcompiler.v1.kernelcompiler.Kernel;
import com.maxeler.maxcompiler.v1.kernelcompiler.KernelParameters;
import com.maxeler.maxcompiler.v1.kernelcompiler.types.base.HWVar;
import com.maxeler.maxcompiler.v1.kernelcompiler.stdlib.core.CounterChain;


public class KernelFunc extends Kernel {

public KernelFunc(KernelParameters parameters) {
super(parameters);
HWVar x =  io.input("x", hwFloat(8, 24));
HWVar c =  io.scalarInput("c", hwFloat(8, 24));
HWVar a1 =  io.scalarInput("a1", hwUInt(32));
HWVar a2 =  io.scalarInput("a2", hwUInt(32));
HWVar a3 =  io.scalarInput("a3", hwUInt(32));
HWVar n1 =  io.scalarInput("n1", hwUInt(32));
HWVar n2 =  io.scalarInput("n2", hwUInt(32));
HWVar n3 =  io.scalarInput("n3", hwUInt(32));
HWVar nx =  io.scalarInput("nx", hwUInt(32));
HWVar ny =  io.scalarInput("ny", hwUInt(32));
HWVar nz =  io.scalarInput("nz", hwUInt(32));

CounterChain chain_i = control.count.makeCounterChain();

HWVar i = chain_i.addCounter(nz, 1);
HWVar j = chain_i.addCounter(ny, 1);
HWVar k = chain_i.addCounter(nx, 1);
HWVar cond = ((((i >= a1) & (i < n1)) & ((j >= a2) & (j < n2))) & ((k >= a3) & (k < n3)));
HWVar func = (((((((c * stream.offset(x, 0)) + (c * stream.offset(x, 1))) + (c * stream.offset(x, -constant.var(1), -1024, 1024))) + (c * stream.offset(x, -nx, -1024, 1024))) + (c * stream.offset(x, nx, -1024, 1024))) + (c * stream.offset(x, (-nx * ny), -1024, 1024))) + (c * stream.offset(x, (nx * ny), -1024, 1024)));

}
}
HWVar res =  io.output("res",hwFloat(8, 24));


